webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/AgglomerateService.scala

```diff
@@ -189,4 +191,78 @@ class AgglomerateService @Inject()(config: DataStoreConfig) extends DataConverte
     } catch {
       case e: Exception => Failure(e.getMessage)
     }
+
+  def largestAgglomerateId(agglomerateFileKey: AgglomerateFileKey): Box[Long] = {
+    val hdfFile = agglomerateFileKey.path(dataBaseDir, agglomerateDir, agglomerateFileExtension).toFile
+
+    tryo {
+      val reader = HDF5FactoryProvider.get.openForReading(hdfFile)
+      reader.`object`().getNumberOfElements("/agglomerate_to_segments_offsets") - 1L
+    }
+  }
+
+  def agglomerateIdsForSegmentIds(agglomerateFileKey: AgglomerateFileKey, segmentIds: List[Long]): Box[List[Long]] = {
+    val cachedAgglomerateFile = agglomerateFileCache.withCache(agglomerateFileKey)(initHDFReader)
+
+    tryo {
+      val agglomerateIds = segmentIds.map { segmentId =>
+        cachedAgglomerateFile.agglomerateIdCache.withCache(ULong(segmentId),
+                                                           cachedAgglomerateFile.reader,
+                                                           cachedAgglomerateFile.dataset)(readHDF)
+      }
+      cachedAgglomerateFile.finishAccess()
+      agglomerateIds
+    }
+
+  }
+
+  def generateAgglomerateGraph(agglomerateFileKey: AgglomerateFileKey, agglomerateId: Long): Box[AgglomerateGraph] =
+    tryo {
+      val hdfFile = agglomerateFileKey.path(dataBaseDir, agglomerateDir, agglomerateFileExtension).toFile
+
+      val reader = HDF5FactoryProvider.get.openForReading(hdfFile)
+
+      val positionsRange: Array[Long] =
+        reader.uint64().readArrayBlockWithOffset("/agglomerate_to_segments_offsets", 2, agglomerateId)
+      val edgesRange: Array[Long] =
+        reader.uint64().readArrayBlockWithOffset("/agglomerate_to_edges_offsets", 2, agglomerateId)
+
+      logger.info(s"positionsRange: ${positionsRange(0)} to ${positionsRange(1)}, agglomerateId: $agglomerateId")
+
+      val nodeCount = positionsRange(1) - positionsRange(0)
+      val edgeCount = edgesRange(1) - edgesRange(0)
+      val edgeLimit = config.Datastore.AgglomerateSkeleton.maxEdges
+      if (nodeCount > edgeLimit) {
+        throw new Exception(s"Agglomerate has too many nodes ($nodeCount > $edgeLimit)")
+      }
```

Comment:
We cannot assume that there is at least an edge per node (e.g. thinking about the current use-case with the sparse agglomerates and adding all other segments as their own agglomerates). So I'd find it confusing to change maxEdges because of the number of nodes. Maybe just rename the config entry to `nodeAndEdgeLimit` or `maxNodesAndEdges`?

Reply:
Makes sense but I’d vote to defer this, in order not to make the migrations needed for deploying this PR more complicated, especially since this problem was not introduced here but already existed.

Reply:
Fair :+1: